File: server/datasets/python_source-code/Multiple-E1/type3_stack_1_grid_dfs.py
def numIslands(grid):
    if not grid:
        return 0
    count = 0
    for r in range(len(grid)):
        for c in range(len(grid[0])):
            if grid[r][c] == '1':
                count += 1
                stack = [(r, c)]
                while stack:
                    curr_r, curr_c = stack.pop()
                    if 0 <= curr_r < len(grid) and 0 <= curr_c < len(grid[0]) and grid[curr_r][curr_c] == '1':
                        grid[curr_r][curr_c] = '0'
                        stack.append((curr_r + 1, curr_c))
                        stack.append((curr_r - 1, curr_c))
                        stack.append((curr_r, curr_c + 1))
                        stack.append((curr_r, curr_c - 1))
    return count

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_e1/type3_stack_1_grid_dfs.java
import java.util.Stack;

public class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }
        int count = 0;
        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[0].length; c++) {
                if (grid[r][c] == '1') {
                    count++;
                    Stack<int[]> stack = new Stack<>();
                    stack.push(new int[]{r, c});
                    while (!stack.isEmpty()) {
                        int[] curr = stack.pop();
                        int curr_r = curr[0];
                        int curr_c = curr[1];
                        if (curr_r >= 0 && curr_r < grid.length && curr_c >= 0 && curr_c < grid[0].length && grid[curr_r][curr_c] == '1') {
                            grid[curr_r][curr_c] = '0';
                            stack.push(new int[]{curr_r + 1, curr_c});
                            stack.push(new int[]{curr_r - 1, curr_c});
                            stack.push(new int[]{curr_r, curr_c + 1});
                            stack.push(new int[]{curr_r, curr_c - 1});
                        }
                    }
                }
            }
        }
        return count;
    }
}
